effectful/handlers/llm/synthesis.py

```diff
@@ -19,82 +28,408 @@ def __init__(self, message, code=None):
-        lines = code.splitlines(keepends=True)
-        filename = f"<generated-{hash(code)}>"
+        Executes the module code and returns the named function.
+        The module code becomes the function's lexical context,
+        optionally augmented with provided context.
+        """
+        func_name = vl.function_name
+        module_code = textwrap.dedent(vl.module_code).strip()
 
-        # register into linecache
-        linecache.cache[filename] = (len(source_code), None, lines, filename)
+        cls._decode_counter += 1
+        filename = f"<synthesized:{func_name}:{cls._decode_counter}>"
+        lines = module_code.splitlines(keepends=True)
+        # Ensure last line has newline for linecache
+        if lines and not lines[-1].endswith("\n"):
+            lines[-1] += "\n"
+        linecache.cache[filename] = (
+            len(module_code),
+            None,
+            lines,
+            filename,
+        )
+
+        # Start with provided context or empty dict
+        # Include collections module for type hints in synthesized code
+        exec_globals: dict[str, typing.Any] = {"collections": collections}
+        if context:
+            exec_globals.update(context)
 
-        # TODO: support injecting lexical context for synthesized code
-        gs: dict = {}
         try:
-            code_obj = compile(source_code, filename, "exec")
-            exec(code_obj, gs)
+            code_obj = compile(module_code, filename, "exec")
+            exec(code_obj, exec_globals)
+        except SyntaxError as exc:
+            raise SynthesisError(
+                f"Syntax error in generated code: {exc}", module_code
+            ) from exc
         except Exception as exc:
-            raise SynthesisError("evaluation failed", content) from exc
+            raise SynthesisError(f"Evaluation failed: {exc!r}", module_code) from exc
+
+        if func_name not in exec_globals:
+            raise SynthesisError(
+                f"Function '{func_name}' not found after execution. "
+                f"Available names: {[k for k in exec_globals.keys() if not k.startswith('_')]}",
+                module_code,
+            )
+
+        func = exec_globals[func_name]
+        # Also attach source code directly for convenience
+        func.__source__ = module_code
+        func.__synthesized__ = vl
+        return func
+
+    @classmethod
+    def serialize(cls, vl: SynthesizedFunction) -> list[OpenAIMessageContentListBlock]:
+        return [{"type": "text", "text": vl.model_dump_json()}]
+
+
+@type_to_encodable_type.register(LexicalContext)
+class EncodableLexicalContext(
+    EncodableAs[LexicalContext, str],
+):
+    """Encodes LexicalContext to Python source code representation.
+
+    encode: Extracts source code/signatures for types, functions, and values.
+    decode: Executes source code to reconstruct a lexical context.
+    """
+
+    t = str
+
+    @classmethod
+    def encode(cls, context: LexicalContext) -> str:
```

Comment:
I would expect `EncodableLexicalContext.encode` to recursively call `type_to_encodable_type`, not implement custom serialization logic for a bunch of other types internally.